File: win/ServerRegistry.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace MakeYourChoice
{
    /// <summary>
    /// EXPERIMENT — remembers the real DBD game-server IPs we've actually connected to, grouped by
    /// AWS region code, so the active beacon (LiveProbe) has concrete targets to probe even when the
    /// game isn't running. Fed by the traffic sniffer; persisted to disk across launches.
    ///
    /// This is the "work in just the app without connecting" piece: the more you've played, the
    /// bigger the per-region pool of known servers, so on a cold launch we can probe that pool and
    /// tell whether the fleet is up — no need to be in a match.
    ///
    /// PERSISTENCE: plain-text, one record per line ("regionCode|ip|port|lastSeenUnix"). We do NOT
    /// use System.Text.Json here because the app is published with PublishTrimmed, which breaks
    /// reflection-based JSON (de)serialization at runtime — that silently wiped the registry before.
    /// </summary>
    public sealed class ServerRegistry
    {
        public sealed class Entry
        {
            public string Ip { get; set; }
            public int Port { get; set; }
            public long LastSeenUnix { get; set; }
            public long LastLiveUnix { get; set; }   // last time a probe confirmed this endpoint live
            // Recency used for ranking: the most recent of "seen in a real connection" or "probed live".
            public long Recency => Math.Max(LastSeenUnix, LastLiveUnix);
        }

        // One game-server instance (IP) with its best representative port — the unit the beacon
        // dedups to (probe one port per IP first).
        public readonly struct Instance
        {
            public Instance(string ip, int port, long recency, long lastLive)
            { Ip = ip; Port = port; Recency = recency; LastLiveUnix = lastLive; }
            public string Ip { get; }
            public int Port { get; }
            public long Recency { get; }
            public long LastLiveUnix { get; }
        }

        // Keep a generous history per region so a cold launch has many probe targets even as
        // GameLift churns individual instances.
        private const int MaxPerRegion = 256;

        private readonly object _lock = new();
        private readonly Dictionary<string, Dictionary<string, Entry>> _byRegion = new(); // region -> (ip -> entry)
        private readonly string _path;

        public ServerRegistry()
        {
            var dir = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "MakeYourChoice");
            try { Directory.CreateDirectory(dir); } catch { }
            _path = Path.Combine(dir, "known-servers.txt");
            // Shared address book shipped with the app (curated from many users' connections), so a
            // fresh install can probe known DBD subnets without discovering anything itself.
            LoadFile(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "servers-seed.txt"));
            Load(); // user's own learned servers (override/extend the shipped seed)
        }

        public void Record(string regionCode, string ip, int port)
        {
            if (string.IsNullOrEmpty(regionCode) || string.IsNullOrEmpty(ip)) return;
            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            lock (_lock)
            {
                if (!_byRegion.TryGetValue(regionCode, out var map))
                    _byRegion[regionCode] = map = new Dictionary<string, Entry>();
                var key = ip + ":" + port;
                if (map.TryGetValue(key, out var e)) e.LastSeenUnix = now; // preserve LastLiveUnix
                else map[key] = new Entry { Ip = ip, Port = port, LastSeenUnix = now };
                Trim(map);
            }
            Save();
        }

        /// <summary>Mark an endpoint confirmed-live (a probe got a DBD challenge). Powers reliability
        /// ranking and self-pruning. Not persisted on every call — flushed opportunistically.</summary>
        public void MarkLive(string regionCode, string ip, int port)
        {
            if (string.IsNullOrEmpty(regionCode) || string.IsNullOrEmpty(ip)) return;
            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            lock (_lock)
            {
                if (!_byRegion.TryGetValue(regionCode, out var map))
                    _byRegion[regionCode] = map = new Dictionary<string, Entry>();
                var key = ip + ":" + port;
                if (map.TryGetValue(key, out var e)) { e.LastLiveUnix = now; }
                else map[key] = new Entry { Ip = ip, Port = port, LastSeenUnix = now, LastLiveUnix = now };
            }
        }

        /// <summary>Persist current state (call after a batch of MarkLive updates).</summary>
        public void Flush() => Save();

        /// <summary>
        /// Distinct instances (one per IP) for a region, newest-first by recency, with each IP's best
        /// (most recent) port as the representative. This is what the beacon probes — one port per IP.
        /// </summary>
        public List<Instance> GetInstancesRanked(string regionCode, int max)
        {
            lock (_lock)
            {
                if (regionCode == null || !_byRegion.TryGetValue(regionCode, out var map))
                    return new List<Instance>();
                var byIp = new Dictionary<string, Entry>();
                foreach (var e in map.Values)
                    if (!byIp.TryGetValue(e.Ip, out var best) || e.Recency > best.Recency)
                        byIp[e.Ip] = e;
                return byIp.Values
                    .OrderByDescending(e => e.Recency)
                    .Take(max)
                    .Select(e => new Instance(e.Ip, e.Port, e.Recency, e.LastLiveUnix))
                    .ToList();
            }
        }

        /// <summary>Drop endpoints that haven't been seen or confirmed live in maxAgeDays AND have
        /// never been probed live — keeps the pool lean so we probe fewer, higher-quality targets.</summary>
        public void Prune(int maxAgeDays = 14)
        {
            long cutoff = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - (long)maxAgeDays * 86400;
            bool changed = false;
            lock (_lock)
            {
                foreach (var map in _byRegion.Values)
                    foreach (var e in map.Values.Where(e => e.Recency < cutoff && e.LastLiveUnix == 0).ToList())
                    { map.Remove(e.Ip + ":" + e.Port); changed = true; }
            }
            if (changed) Save();
        }

        private static void Trim(Dictionary<string, Entry> map)
        {
            if (map.Count <= MaxPerRegion) return;
            foreach (var old in map.Values.OrderBy(e => e.LastSeenUnix).Take(map.Count - MaxPerRegion).ToList())
                map.Remove(old.Ip + ":" + old.Port);
        }

        /// <summary>Most-recently-seen server endpoints for a region (newest first).</summary>
        public List<Entry> GetCandidates(string regionCode, int max)
        {
            lock (_lock)
            {
                if (regionCode == null || !_byRegion.TryGetValue(regionCode, out var map))
                    return new List<Entry>();
                return map.Values.OrderByDescending(e => e.LastSeenUnix).Take(max).ToList();
            }
        }

        public int RegionCount { get { lock (_lock) return _byRegion.Count; } }
        public int TotalServers { get { lock (_lock) return _byRegion.Values.Sum(m => m.Count); } }
        public int CountFor(string regionCode)
        {
            lock (_lock) return _byRegion.TryGetValue(regionCode, out var m) ? m.Count : 0;
        }

        /// <summary>Distinct /24 prefixes (e.g. "18.225.57") seen for a region — the subnets to sweep.</summary>
        public List<string> GetSubnets24(string regionCode)
        {
            lock (_lock)
            {
                if (regionCode == null || !_byRegion.TryGetValue(regionCode, out var map)) return new List<string>();
                var set = new HashSet<string>();
                foreach (var e in map.Values)
                {
                    var dot = e.Ip.LastIndexOf('.');
                    if (dot > 0) set.Add(e.Ip.Substring(0, dot));
                }
                return set.ToList();
            }
        }

        /// <summary>Distinct game ports DBD has used in a region — the ports to probe when sweeping.</summary>
        public List<int> GetKnownPorts(string regionCode)
        {
            lock (_lock)
            {
                if (regionCode == null || !_byRegion.TryGetValue(regionCode, out var map)) return new List<int>();
                return map.Values.Select(e => e.Port).Distinct().ToList();
            }
        }

        private void Load() => LoadFile(_path);

        // Merge a "regionCode|ip|port|lastSeenUnix" file into the registry (newer LastSeen wins).
        private void LoadFile(string path)
        {
            try
            {
                if (!File.Exists(path)) return;
                lock (_lock)
                {
                    foreach (var raw in File.ReadAllLines(path))
                    {
                        var line = raw.Trim();
                        if (line.Length == 0 || line[0] == '#') continue;
                        var p = line.Split('|');
                        if (p.Length < 4) continue;
                        var region = p[0].Trim();
                        var ip = p[1].Trim();
                        if (region.Length == 0 || ip.Length == 0) continue;
                        if (!int.TryParse(p[2], out var port)) continue;
                        if (!long.TryParse(p[3], out var seen)) seen = 0;
                        long live = (p.Length >= 5 && long.TryParse(p[4], out var lv)) ? lv : 0; // optional 5th field
                        if (!_byRegion.TryGetValue(region, out var map))
                            _byRegion[region] = map = new Dictionary<string, Entry>();
                        var key = ip + ":" + port;
                        if (!map.TryGetValue(key, out var existing) || existing.LastSeenUnix < seen)
                            map[key] = new Entry { Ip = ip, Port = port, LastSeenUnix = seen, LastLiveUnix = live };
                    }
                }
            }
            catch { /* corrupt/missing -> ignore */ }
        }

        private void Save()
        {
            try
            {
                var sb = new System.Text.StringBuilder();
                sb.AppendLine("# DBD known servers — regionCode|ip|port|lastSeenUnix|lastLiveUnix");
                lock (_lock)
                {
                    foreach (var region in _byRegion)
                        foreach (var e in region.Value.Values)
                            sb.Append(region.Key).Append('|').Append(e.Ip).Append('|')
                              .Append(e.Port).Append('|').Append(e.LastSeenUnix).Append('|')
                              .Append(e.LastLiveUnix).Append('\n');
                }
                File.WriteAllText(_path, sb.ToString());
            }
            catch { /* best effort */ }
        }
    }

    /// <summary>
    /// EXPERIMENT — appends active-beacon probe results to a plain-text log the user can read or send
    /// back for analysis (cross-checked against a Wireshark capture).
    /// </summary>
    public static class BeaconLog
    {
        // Off by default; the Debug toggle (Experimental settings) turns it on at runtime so you can
        // tune the beacon from beacon-log.txt without a special build.
        public static volatile bool Enabled = false;

        private static readonly object _lock = new();
        public static readonly string Path = System.IO.Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "MakeYourChoice", "beacon-log.txt");

        public static void Write(string line)
        {
            if (!Enabled) return;
            try
            {
                lock (_lock)
                {
                    System.IO.Directory.CreateDirectory(System.IO.Path.GetDirectoryName(Path));
                    File.AppendAllText(Path, $"{DateTime.Now:yyyy-MM-dd HH:mm:ss}  {line}{Environment.NewLine}");
                }
            }
            catch { /* best effort */ }
        }
    }
}

File: linux/src/server_registry.rs
// Remembers real DBD game-server endpoints per AWS region (Linux port of ServerRegistry.cs).
// Fed by the traffic sniffer; persisted across launches; ships with a shared address book so the
// beacon can probe known servers without the user connecting first. Thread-safe so the sniffer's
// async tasks and the beacon timer can both use it.

use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

// Shared address book compiled into the binary (curated from real connections).
const SEED: &str = include_str!("../servers-seed.txt");
const MAX_PER_REGION: usize = 256;

#[derive(Clone)]
pub struct Entry {
    pub ip: String,
    pub port: u16,
    pub last_seen: u64,
    pub last_live: u64, // last time a probe confirmed this endpoint live
}

impl Entry {
    fn recency(&self) -> u64 {
        self.last_seen.max(self.last_live)
    }
}

// One game-server instance (IP) with its best representative port — the beacon dedups to this.
#[derive(Clone)]
pub struct Instance {
    pub ip: String,
    pub port: u16,
}

pub struct ServerRegistry {
    // region -> ("ip:port" -> entry)
    by_region: Mutex<HashMap<String, HashMap<String, Entry>>>,
    path: PathBuf,
}

fn now_unix() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

impl ServerRegistry {
    pub fn new() -> Self {
        let dir = dirs::config_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join("make-your-choice");
        let _ = fs::create_dir_all(&dir);
        let reg = ServerRegistry {
            by_region: Mutex::new(HashMap::new()),
            path: dir.join("known-servers.txt"),
        };
        reg.merge_text(SEED); // shipped seed
        if let Ok(text) = fs::read_to_string(&reg.path) {
            reg.merge_text(&text); // user's learned servers
        }
        reg
    }

    fn merge_text(&self, text: &str) {
        let mut map = self.by_region.lock().unwrap();
        for raw in text.lines() {
            let line = raw.trim();
            if line.is_empty() || line.starts_with('#') {
                continue;
            }
            let parts: Vec<&str> = line.split('|').collect();
            if parts.len() < 4 {
                continue;
            }
            let region = parts[0].trim();
            let ip = parts[1].trim();
            let port: u16 = match parts[2].trim().parse() {
                Ok(p) => p,
                Err(_) => continue,
            };
            let seen: u64 = parts[3].trim().parse().unwrap_or(0);
            let live: u64 = parts.get(4).and_then(|s| s.trim().parse().ok()).unwrap_or(0);
            if region.is_empty() || ip.is_empty() {
                continue;
            }
            let entry = Entry { ip: ip.to_string(), port, last_seen: seen, last_live: live };
            let inner = map.entry(region.to_string()).or_default();
            let key = format!("{}:{}", ip, port);
            match inner.get(&key) {
                Some(e) if e.last_seen >= seen => {}
                _ => {
                    inner.insert(key, entry);
                }
            }
        }
    }

    pub fn record(&self, region_code: &str, ip: &str, port: u16) {
        if region_code.is_empty() || ip.is_empty() {
            return;
        }
        {
            let mut map = self.by_region.lock().unwrap();
            let inner = map.entry(region_code.to_string()).or_default();
            let key = format!("{}:{}", ip, port);
            match inner.get_mut(&key) {
                Some(e) => e.last_seen = now_unix(), // preserve last_live
                None => {
                    inner.insert(key, Entry { ip: ip.to_string(), port, last_seen: now_unix(), last_live: 0 });
                }
            }
            if inner.len() > MAX_PER_REGION {
                let mut items: Vec<(String, u64)> =
                    inner.iter().map(|(k, e)| (k.clone(), e.last_seen)).collect();
                items.sort_by_key(|(_, t)| *t);
                for (k, _) in items.into_iter().take(inner.len() - MAX_PER_REGION) {
                    inner.remove(&k);
                }
            }
        }
        self.save();
    }

    /// Most-recently-seen endpoints for a region (newest first).
    pub fn candidates(&self, region_code: &str, max: usize) -> Vec<(String, u16)> {
        let map = self.by_region.lock().unwrap();
        match map.get(region_code) {
            None => Vec::new(),
            Some(inner) => {
                let mut v: Vec<&Entry> = inner.values().collect();
                v.sort_by(|a, b| b.last_seen.cmp(&a.last_seen));
                v.into_iter().take(max).map(|e| (e.ip.clone(), e.port)).collect()
            }
        }
    }

    /// Mark an endpoint confirmed-live (a probe got a DBD challenge). Powers reliability ranking and
    /// self-pruning. Not persisted on every call — call flush() after a batch.
    pub fn mark_live(&self, region_code: &str, ip: &str, port: u16) {
        if region_code.is_empty() || ip.is_empty() {
            return;
        }
        let mut map = self.by_region.lock().unwrap();
        let inner = map.entry(region_code.to_string()).or_default();
        let key = format!("{}:{}", ip, port);
        let now = now_unix();
        match inner.get_mut(&key) {
            Some(e) => e.last_live = now,
            None => {
                inner.insert(key, Entry { ip: ip.to_string(), port, last_seen: now, last_live: now });
            }
        }
    }

    /// Distinct instances (one per IP) for a region, newest-first by recency, each IP's best
    /// (most recent) port as representative — the beacon probes one port per IP.
    pub fn instances_ranked(&self, region_code: &str, max: usize) -> Vec<Instance> {
        let map = self.by_region.lock().unwrap();
        let inner = match map.get(region_code) {
            Some(m) => m,
            None => return Vec::new(),
        };
        let mut by_ip: HashMap<&str, &Entry> = HashMap::new();
        for e in inner.values() {
            match by_ip.get(e.ip.as_str()) {
                Some(best) if best.recency() >= e.recency() => {}
                _ => {
                    by_ip.insert(e.ip.as_str(), e);
                }
            }
        }
        let mut v: Vec<&Entry> = by_ip.values().copied().collect();
        v.sort_by(|a, b| b.recency().cmp(&a.recency()));
        v.into_iter().take(max).map(|e| Instance { ip: e.ip.clone(), port: e.port }).collect()
    }

    /// Drop endpoints not seen/live within max_age_days that were never probed live — keeps the pool
    /// lean so the beacon probes fewer, higher-quality targets.
    pub fn prune(&self, max_age_days: u64) {
        let cutoff = now_unix().saturating_sub(max_age_days * 86400);
        let mut changed = false;
        {
            let mut map = self.by_region.lock().unwrap();
            for inner in map.values_mut() {
                let dead: Vec<String> = inner
                    .iter()
                    .filter(|(_, e)| e.recency() < cutoff && e.last_live == 0)
                    .map(|(k, _)| k.clone())
                    .collect();
                for k in dead {
                    inner.remove(&k);
                    changed = true;
                }
            }
        }
        if changed {
            self.save();
        }
    }

    /// Persist current state (call after a batch of mark_live updates).
    pub fn flush(&self) {
        self.save();
    }

    pub fn count_for(&self, region_code: &str) -> usize {
        self.by_region.lock().unwrap().get(region_code).map(|m| m.len()).unwrap_or(0)
    }

    pub fn total(&self) -> usize {
        self.by_region.lock().unwrap().values().map(|m| m.len()).sum()
    }

    fn save(&self) {
        let map = self.by_region.lock().unwrap();
        let mut out = String::from("# DBD known servers — regionCode|ip|port|lastSeenUnix|lastLiveUnix\n");
        for (region, inner) in map.iter() {
            for e in inner.values() {
                out.push_str(&format!("{}|{}|{}|{}|{}\n", region, e.ip, e.port, e.last_seen, e.last_live));
            }
        }
        let _ = fs::write(&self.path, out);
    }
}
